microsoft/iot-cardboard-js
# ResourcePicker component and changes in EnvironmentPicker

### Summary of changes 🔍 
[Azure resource loader for storage container dropdown.docx](https://microsoft.sharepoint.com/:w:/r/teams/Azure_IoT/IoTPlat/Shared%20Documents/Digital%20Twins/36.%20Digital%20Twins%20-%20ExACT%20Team/3D%20Scenes/DevDocs%20(Notion%20cont.)/Feature%20reviews/Azure%20resource%20loader%20for%20storage%20container%20dropdown.docx?d=w2bb53e67bb4f4e258170db431920b3e5&csf=1&web=1&e=P3N0Yv)
[ADO item](https://msazure.visualstudio.com/One/_sprints/taskboard/ADT%20ExACT/One/Custom/IoT/2207?workitem=14600665)

- Added `ResourcePicker` standalone component to pull any type of Azure Resources with a given set of required permissions
   - Added **accountId** to the storage secret file, so please add your own storage account id in your user secret file to test the ResourcePicker with Storage Blob Container resource type. It is required to pull containers more efficiently using **searchParams**, rather than making thousands of calls by pulling all the subscriptions -> storage accounts in all those subscriptions -> storage containers in all those storage accounts -> then checking role assignments for all those?!!
   - Added **take** search parameter to limit the number of resources to check the role definitions for to limit the number of requests being made not to crash the page. The downsize of it is that, when we take first "x" number of resources and check the required permissions against those, the ones that we have access may not be in that batch. _**Please feel free to give any feedback on this.**_ 
- Added `getResourcesByPermissions` method to **AzureManagementAdapter** and revised already existing methods there to clean up the code.
- Consumed ResourcePicker components in the `EnvironmentPicker` (1 for ADT Instances, 2 for Azure Storage Containers)
   - **Using 2 step resource picking for Azure Storage Blob Containers** since there is no other efficient way of fetching containers besides **selecting the storage account first**. Otherwise, we have to make thousands of calls depending on the number of subscriptions and storage accounts of the logged in user. For this purpose, added the storage accounts key to the localstorage to get the account id initially to fetch the containers on mount, otherwise the second resource picker is waiting for the storage accounts resource picker finish and get the id from its results to refetch the containers based on that id.

![image](https://user-images.githubusercontent.com/45217314/177049559-53750f40-f5e6-4a1c-834f-3916e0520691.png)
![image](https://user-images.githubusercontent.com/45217314/177168820-2af81d56-e2f5-4413-a077-61b902074afe.png)
![image](https://user-images.githubusercontent.com/45217314/177169387-29e8da85-9806-4747-aba2-94578e1364b5.png)
<img width="865" alt="image" src="https://user-images.githubusercontent.com/45217314/182250276-fc9c111b-8d12-4354-bae9-3511d1ef9a0a.png">


### Testing 🧪
- Make sure you use first party app secrets (clientId and redirectUri) to use Azure Management API
- Add your own storage account id in your user secret file's "storage" section (see the placeholder file) to test the ResourcePicker with Storage Blob Container resource type
- Use ResourcePicker component's stories and EnvironmentPicker in the ADT3DScenePage
- Enable `debugLogging` in AzureManagementAdapter to see the logs (number of fetched resources etc.)

### Checklist ✔️
